fw/rbcx-coprocessor/src/UltrasoundController.cpp

```diff
@@ -0,0 +1,113 @@
+#include "Dispatcher.hpp"
+
+static const uint16_t pingTimeoutMs = 30;
+static const uint16_t echoTimeoutMs = 200;
+static CoprocStat status;
+static int utsActiveIndex;
+static uint16_t risingEdgeMicros;
+static TaskWrapper<512> utsTask;
+static QueueWrapper<int, 16> trigQueue;
+
+static void enqueueStatus(int utsIndex, uint16_t microseconds) {
+    status = CoprocStat_init_default;
+    status.which_payload = CoprocStat_ultrasoundStat_tag;
+    status.payload.ultrasoundStat.utsIndex = utsIndex;
+    status.payload.ultrasoundStat.roundtripMicrosecs = microseconds;
+    dispatcherEnqueueStatus(status);
+}
+
+void ultrasoundInit() {
+    LL_TIM_InitTypeDef timInit;
+    LL_TIM_StructInit(&timInit);
+    auto apb1TimClk = 2 * HAL_RCC_GetPCLK1Freq();
+
+    timInit.Autoreload = pingTimeoutMs * 1000;
+    timInit.Prescaler = apb1TimClk / 1'000'000; // 1 us ticks
+    LL_TIM_Init(utsTimer, &timInit);
+    LL_TIM_SetOnePulseMode(utsTimer, LL_TIM_ONEPULSEMODE_SINGLE);
+    LL_TIM_SetUpdateSource(utsTimer, LL_TIM_UPDATESOURCE_COUNTER);
+    LL_TIM_ClearFlag_UPDATE(utsTimer);
+
+    LL_EXTI_DisableIT_0_31(utsEchoPins.second);
+
+    trigQueue.create();
+    utsTask.start("ultrasound", 2, []() {
+        while (true) {
+            int utsIndex;
+            trigQueue.pop_front(utsIndex);
+
+            // Wait for potential ECHO high
+            while (pinRead(utsEchoPin[utsIndex])) {
+                vTaskDelay(pdMS_TO_TICKS(5));
+            }
+            if (pinRead(utsEchoPin[utsIndex])) {
+                DEBUG("ECHO %d hanging high", utsIndex);
+                continue;
+            }
+
+            LL_TIM_GenerateEvent_UPDATE(utsTimer);
+            LL_TIM_ClearFlag_UPDATE(utsTimer);
+
+            // Perform TRIG pulse
+            taskENTER_CRITICAL();
+            LL_TIM_EnableCounter(utsTimer);
+            pinWrite(utsTrigPin[utsIndex], true);
+            while (LL_TIM_GetCounter(utsTimer) < 10
+                && !LL_TIM_IsActiveFlag_UPDATE(utsTimer)) { // 10 us pulse
+            }
+            pinWrite(utsTrigPin[utsIndex], false);
+            taskEXIT_CRITICAL();
+
+            // Wait for ECHO measurement from ISR
+            utsActiveIndex = utsIndex;
+            LL_EXTI_EnableIT_0_31(utsEchoPin[utsIndex].second);
+            uint32_t micros;
+            auto success = xTaskNotifyWait(
+                0U, ~0U, &micros, pdMS_TO_TICKS(pingTimeoutMs));
+            LL_EXTI_DisableIT_0_31(utsEchoPin[utsIndex].second);
+            if (success) {
+                enqueueStatus(utsIndex, micros);
+            } else {
+                enqueueStatus(utsIndex, 0);
+            }
+        }
+    });
+}
+
+void ultrasoundDispatch(const CoprocReq_UltrasoundReq& request) {
+    if (request.utsIndex > 3) {
+        DEBUG("Ultrasound index %lu out of range", request.utsIndex);
+        return;
+    }
+
+    switch (request.which_utsCmd) {
+    case CoprocReq_UltrasoundReq_singlePing_tag:
+        trigQueue.push_back(request.utsIndex, 0);
```

Comment:
Když se request nevleze do queue, mohl by něco poslat zpátky, nějakou chybovou hodnotu.

Reply:
Jako mohlo by být, ale v takových případech se jedná o selhání v použití. Dám tam DEBUG